README
Please see the current maintainer's repository at
https://github.com/h0tw1r3/pam_shield

pam_shield by Walter de Jong <walter@heiho.net> and Jonathan Niehof
<jtniehof@gmail.com>, copyright 2007-2012.

pam_shield COMES WITH NO WARRANTY. pam_shield IS FREE SOFTWARE.
pam_shield is distributed under terms described in the GNU General Public
License.

See the INSTALL file for information on how to install pam_shield.


pam_shield is a PAM module that uses iptables or null-routing to lock out
script kiddies that probe your computer for open logins and/or easy guessable
passwords. pam_shield is meant as an aid to protect public computers on the
open internet.

Everybody knows it is unwise to leave computers largely unprotected
connected to the internet. However, there are cases in which this is still
common practice. For exampe, academic sites with hundreds of users often
have a policy of allowing logins from over the world. They are under
constant attack by "kiddies" trying to break in to the system by
password guessing. pam_shield aims to detect and block these "kiddies".


(Not So) Random Remarks
-----------------------
 * pam_shield is a PAM (Pluggable Authentication Module). When used
   inappropriately, your system might be at risk. Use with care.

 * pam_shield blocks IPs. This means that when it blocks a multi-user
   system, it blocks all users from that system.
   For example, it may happen that an attacker is performing his
   attack from a university system, from which many students connect.
   By blocking the attacker, all students get blocked as well. This should
   be no problem, but you should be aware that this can happen.

 * pam_shield works by counting login attempts coming from a remote host
   during a period of time. If there are too many attempts, it triggers
   and blocks the remote host.

 * To block and unblock IPs, pam_shield runs the shield-trigger script.
   By default, it uses null-routing to block hosts.
   A script for using iptables is also provided, but you should customize
   this script to fit your situation if you decide to use it.

 * similar tools are daemon_shield and BlockHosts, which work by scanning
   system logs. pam_shield works with PAM and a gdbm database.

 * pam_shield is by no means THE solution for all your security problems.
   Always remain on guard.


See Also
--------
 * iptables homepage: http://www.netfilter.org/

 * Linux PAM documentation:
   http://www.kernel.org/pub/linux/libs/pam/Linux-PAM-html/Linux-PAM_SAG.html

 * fail2ban: http://www.fail2ban.org/

 * daemon_shield: https://sourceforge.net/projects/daemonshield/

 * BlockHosts: http://www.aczoom.com/cms/blockhosts/


History
-------
2007 Walter de Jong created pam_shield.
2010 Walter and Jonathan Niehof started co-maintaining.
2012 Jonathan became the primary maintainer.
2022 Jeffrey Clark became the primary maintainer.


Changelog
30 Aug 2022:
- release 0.9.7
- fix gdbm_option compiler warnings
- format database list output as json
- parse config file first to support runtime overrides
- new optional argument to remove single IP from database
- new primary maintainer Jeffrey Clark

3 Jun 2012:
- release 0.9.6
- Change build system to autoconf
- Add man pages (Thanks to Carl Thompson)
- Make shield-trigger-iptables more flexible (Thanks Carl Thompson)
- Make cron job behave if shield-purge disappears from under it
- use iptables to block bad IPs on all interfaces, not just eth0
- add optional trigger script to use ufw instead of iptables directly
- new primary maintainer Jonathan Niehof

12 Jan 2011:
- release 0.9.5
Bugfix by Heino Gutschmidt:
- shield_purge would delete entries that were still active,
  causing the delete-rule trigger never to be executed

4 Sept 2010:
- release 0.9.4

1 Sept 2010:
- bugfix: treat missing DNS parameters properly
- bugfix: for race of blocking same IP multiple times
- bugfix: check for exit code of run_trigger
- read_config() may return an (syntax) error, but continue anyway
  and try to make the best of it
- use ip blockhole routing
- added lots of explanatory text to INSTALL

20 April 2010: release 0.9.3
Bugfixes by Jonathan Niehof:
- fix memory leak in purger
- fix bug in the code where the purger deletes multiple entries
- when opening the database fails, retry a couple of times

New feature by Jonathan Niehof:
- added option --force to purger to delete all entries


2007: release 0.9.2
Version 0.9.2 was the first public release


pam_shield.c
/*
	pam_shield.c

    pam_shield 0.9.7
    Copyright (C) 2007-2012  Walter de Jong <walter@heiho.net>
    and Jonathan Niehof <jtniehof@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
/*
	pam_shield is a PAM module that uses route/iptables to lock out script kiddies
	that probe your machine for open logins and/or easy guessable passwords.

	You can run this module with

	auth optional	pam_shield.so

	But just make sure it's not the only auth module you run..!
	This module does not do any authentication, it just monitors access.


	(btw, if you don't like my indentation, try setting tabsize to 4)
*/

#include "pam_shield.h"

#define PAM_SM_AUTH		1

#include <security/pam_modules.h>

#include "pam_shield_lib.h"

#pragma GCC visibility push(hidden)

void logmsg(int level, const char *fmt, ...) {
va_list varargs;

	if (level == LOG_DEBUG && !(options & OPT_DEBUG))
		return;

#ifdef LOG_AUTHPRIV
	openlog("PAM-shield", LOG_PID, LOG_AUTHPRIV);
#else
	openlog("PAM-shield", LOG_PID, LOG_AUTH);
#endif

	va_start(varargs, fmt);
	vsyslog(level, fmt, varargs);
	va_end(varargs);

	closelog();
}

/*
	Mind that argv[0] is an argument, not the name of the module
*/
static void get_options(int argc, char **argv) {
int i;

	for(i = 0; i < argc; i++) {
		if (!strcmp(argv[i], "debug")) {
			options |= OPT_DEBUG;
			logmsg(LOG_DEBUG, "logging debug info");
			continue;
		}
		if (!strcmp(argv[i], "use_first_pass"))			/* Thorsten Kukuk sez all modules should accept this argument */
			continue;

		if (!strncmp(argv[i], "conf=", 5)) {
			conffile = argv[i] + 5;
			continue;
		}
		logmsg(LOG_ERR, "unknown argument '%s', ignored", argv[i]);
	}
}

static _pam_shield_db_rec_t *new_db_record(int window_size) {
_pam_shield_db_rec_t *record;
int size;

	if (window_size <= 0) {
		window_size = 1;
		size = sizeof(_pam_shield_db_rec_t);
	} else
		size = sizeof(_pam_shield_db_rec_t) + (window_size-1) * sizeof(time_t);

	if ((record = (_pam_shield_db_rec_t *)malloc(size)) == NULL) {
		logmsg(LOG_CRIT, "new_db_record(): out of memory allocating %d bytes", size);
		return NULL;
	}
	memset(record, 0, size);
	record->max_entries = window_size;
	return record;
}

static void destroy_db_record(_pam_shield_db_rec_t *record) {
	if (record != NULL)
		free(record);
}

/*
	get remote IPs for the rhost

	the return value must be freed with freeaddrinfo()
*/
static struct addrinfo *get_addr_info(char *rhost) {
struct addrinfo hints, *res;
int err;

	memset(&hints, 0, sizeof(struct addrinfo));
	hints.ai_family = PF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;

	if ((err = getaddrinfo(rhost, NULL, &hints, &res)) != 0) {
		logmsg(LOG_ERR, "%s: %s\n", rhost, gai_strerror(err));
		return NULL;
	}
	return res;
}

#pragma GCC visibility pop

/*
	the authenticate function always returns PAM_IGNORE, because this
	module does not really authenticate
*/
PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv) {
char *user, *rhost;
struct passwd *pwd;
unsigned int retry_count;
int suspicious_dns;

	if (init_module())
		return PAM_IGNORE;

	logmsg(LOG_DEBUG, "this is version " PAM_SHIELD_VERSION);

/*
	read_config() may fail (due to syntax errors, etc.), try to make the best of it
	by continuing anyway
*/
	read_config();
	get_options(argc, (char **)argv);

/* get the username */
	if (pam_get_item(pamh, PAM_USER, (const void **)(void *)&user) != PAM_SUCCESS)
		user = NULL;

	if (user != NULL && !*user)
		user = NULL;

	logmsg(LOG_DEBUG, "user %s", (user == NULL) ? "(unknown)" : user);

/* if not blocking all and the user is known, let go */
	if (!(options & OPT_BLOCK_ALL) && user != NULL && (pwd = getpwnam(user)) != NULL) {
		logmsg(LOG_DEBUG, "ignoring known user %s", user);
		deinit_module();
		return PAM_IGNORE;
	}

/* get the remotehost address */
	if (pam_get_item(pamh, PAM_RHOST, (const void **)(void *)&rhost) != PAM_SUCCESS)
		rhost = NULL;

	if (rhost != NULL && !*rhost)
		rhost = NULL;

	logmsg(LOG_DEBUG, "remotehost %s", (rhost == NULL) ? "(unknown)" : rhost);

/*
	if rhost is NULL, pam_shield is probably being used for a local service here
	Because pam_shield only makes sense in a networked environment, bail out now
*/
	if (rhost == NULL) {
		deinit_module();
		return PAM_IGNORE;
	}

/*
	if rhost is completely numeric, then it has no DNS entry
*/
	suspicious_dns = 0;
	if (strspn(rhost, "0123456789.") == strlen(rhost)
		|| strspn(rhost, "0123456789:abcdefABCDEF") == strlen(rhost)) {
		if (options & OPT_MISSING_DNS)
			logmsg(LOG_DEBUG, "missing DNS entry for %s (allowed)", rhost);
		else {
			logmsg(LOG_DEBUG, "missing DNS entry for %s (denied)", rhost);
			suspicious_dns = 1;
		}
	} else {
/*
	see if this rhost is whitelisted
*/
		if (match_name_list(rhost)) {
			deinit_module();
			return PAM_IGNORE;
		}
	}
	do {
		struct addrinfo *addr_info, *addr_p;
		unsigned char addr_family;
		char ipbuf[INET6_ADDRSTRLEN], *saddr;
		_pam_shield_db_rec_t *record;
		datum key, data;
		int whitelisted;

		if ((addr_info = get_addr_info(rhost)) == NULL) {		/* missing reverse DNS entry */
			if (options & OPT_MISSING_REVERSE)
				logmsg(LOG_DEBUG, "missing reverse DNS entry for %s (allowed)", rhost);
			else {
				logmsg(LOG_DEBUG, "missing reverse DNS entry for %s (denied)", rhost);
				suspicious_dns = 1;
			}
		}
/* for every address that this host is known for, check for whitelist entry */
		for(addr_p = addr_info; addr_p != NULL; addr_p = addr_p->ai_next) {
			whitelisted = 0;
			switch(addr_p->ai_family) {
				case PF_INET:
					saddr = (char *)&((struct sockaddr_in *)(addr_p->ai_addr))->sin_addr.s_addr;

					if (match_ipv4_list((unsigned char *)saddr)) {
						logmsg(LOG_DEBUG, "remoteip %s (whitelisted)", inet_ntop(AF_INET, (char *)&((struct sockaddr_in *)(addr_p->ai_addr))->sin_addr, ipbuf, sizeof(ipbuf)));
						whitelisted = 1;
					} else
						logmsg(LOG_DEBUG, "remoteip %s", inet_ntop(AF_INET, (char *)&((struct sockaddr_in *)(addr_p->ai_addr))->sin_addr, ipbuf, sizeof(ipbuf)));
					break;
				
				case PF_INET6:
					saddr = (char *)&((struct sockaddr_in6 *)(addr_p->ai_addr))->sin6_addr.s6_addr;

					if (match_ipv6_list((unsigned char *)saddr)) {
						logmsg(LOG_DEBUG, "remoteip %s (whitelisted)", inet_ntop(AF_INET6, (char *)&((struct sockaddr_in6 *)(addr_p->ai_addr))->sin6_addr, ipbuf, sizeof(ipbuf)));
						whitelisted = 1;
					} else
						logmsg(LOG_DEBUG, "remoteip %s", inet_ntop(AF_INET6, (char *)&((struct sockaddr_in6 *)(addr_p->ai_addr))->sin6_addr, ipbuf, sizeof(ipbuf)));
					break;

				default:
					logmsg(LOG_DEBUG, "remoteip unknown (not IP)");

					freeaddrinfo(addr_info);
					deinit_module();
					return (suspicious_dns) ? PAM_AUTH_ERR : PAM_IGNORE;
			}
/* host is whitelisted by an allow line in the config file, so exit */
			if (whitelisted) {
				freeaddrinfo(addr_info);
				deinit_module();
				return PAM_IGNORE;
			}
		}
/* open the database */
		retry_count=0;
		while ((dbf = gdbm_open(dbfile, 512, GDBM_WRCREAT, (mode_t)0600, fatal_func)) == NULL) {
			if (gdbm_errno != GDBM_CANT_BE_WRITER || retry_count>500) {
				logmsg(LOG_ERR, "failed to open gdbm file '%s' : %s", dbfile,
				       gdbm_strerror(gdbm_errno));
				freeaddrinfo(addr_info);
				deinit_module();
				return (suspicious_dns) ? PAM_AUTH_ERR : PAM_IGNORE;
			}
			logmsg(LOG_DEBUG,"waiting to open db, try %d",retry_count);
			usleep(1000);
			retry_count++;
		}
/* for every address that this host is known for, check the database */
		for(addr_p = addr_info; addr_p != NULL; addr_p = addr_p->ai_next) {
			whitelisted = 0;
			switch(addr_p->ai_family) {
				case PF_INET:
					addr_family = PAM_SHIELD_ADDR_IPV4;
					key.dptr = saddr = (char *)&((struct sockaddr_in *)(addr_p->ai_addr))->sin_addr.s_addr;
					key.dsize = sizeof(struct in_addr);
					break;

				case PF_INET6:
					addr_family = PAM_SHIELD_ADDR_IPV6;
					key.dptr = saddr = (char *)&((struct sockaddr_in6 *)(addr_p->ai_addr))->sin6_addr.s6_addr;
					key.dsize = sizeof(struct in6_addr);
					break;

				default:
					addr_family = -1;
					key.dptr = saddr = NULL;
					key.dsize = 0;
			}
			if (key.dptr == NULL)
				continue;

			data = gdbm_fetch(dbf, key);			/* get db record */
			if (data.dptr != NULL) {
				record = (_pam_shield_db_rec_t *)data.dptr;
/*
	Although this code does some expiration, it only does so for "this ip";
	it is still necessary to run an external database cleanup process every
	now and then (eg, from cron.daily)
*/
				expire_record(record);

				if (record->count >= record->max_entries) {		/* shift, so we always log the most recent time */
					memmove(record->timestamps, &record->timestamps[1], (record->max_entries-1)*sizeof(time_t));
					record->count--;
				}
				record->timestamps[record->count++] = this_time;

				logmsg(LOG_DEBUG, "%u times from %s", record->count, rhost);
/*
	too many in the interval, so trigger

	trigger "add" is subject to a race, so try to be smart about it
	and do not add the same block within 20 seconds
*/
				if (record->count >= max_conns && this_time - record->trigger_active > 20
					&& !run_trigger("add", record))
					record->trigger_active = this_time;
			} else {
				if ((record = new_db_record(max_conns)) != NULL) {
					record->addr_family = addr_family;
					memcpy(record->ip.any, saddr, key.dsize);
					record->timestamps[record->count++] = this_time;

					logmsg(LOG_DEBUG, "putting new record in db");

					if (max_conns <= 1) {		/* (maybe) stupid, but possible */
						record->trigger_active = this_time;
						run_trigger("add", record);
					}
				}
			}
			if (record != NULL) {
				data.dptr = (char *)record;
				data.dsize = sizeof(_pam_shield_db_rec_t) + (record->max_entries-1)*sizeof(time_t);

/* key.dptr and key.dsize are still set to saddr and addr_size */

				if (gdbm_store(dbf, key, data, GDBM_REPLACE))
					logmsg(LOG_ERR, "failed to write db record");
			}
			destroy_db_record(record);
		}
		freeaddrinfo(addr_info);
		gdbm_close(dbf);
	} while(0);

	deinit_module();
	return (suspicious_dns) ? PAM_AUTH_ERR : PAM_IGNORE;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv) {
	return PAM_SUCCESS;
}

/* EOB */


pam_shield.h
/*
	pam_shield.h
*/

#ifndef PAM_SHIELD
#define PAM_SHIELD	1

#include <netinet/in.h>

#define PAM_SHIELD_ADDR_IPV4		0
#define PAM_SHIELD_ADDR_IPV6		1

typedef struct {
	unsigned char addr_family;	/* PAM_SHIELD_ADDR_IPV4|PAM_SHIELD_ADDR_IPV6 */
	union {
		struct in_addr in;		/* IPv4 number */
		struct in6_addr in6;	/* IPv6 number */
		char any[1];			/* access to any */
	} ip;

	unsigned int max_entries;	/* number of timestamps */
	unsigned int count;			/* number of auth requests done */
	time_t trigger_active;		/* time the trigger was triggered (needed for expiration) */
	time_t timestamps[1];		/* sliding window of timestamps */
} _pam_shield_db_rec_t;


/*
	the IP list is used for making in-memory whitelists
	(they are not in the database, but in the config file)
*/
typedef struct ip_list_tag ip_list;

struct ip_list_tag {
	union {
		struct in_addr in;
		struct in6_addr in6;
		unsigned char any[1];
	} ip;

	union {
		struct in_addr in;
		struct in6_addr in6;
		unsigned char any[1];
	} mask;

	ip_list *prev, *next;
};

/* whitelisted hosntnames and network names */
typedef struct name_list_tag name_list;

struct name_list_tag {
	name_list *prev, *next;

	char name[1];
};

#endif	/* PAM_SHIELD */

/* EOB */


pam_shield_lib.h
/*
	pam_shield_lib.h

    pam_shield 0.9.7
    Copyright (C) 2007-2012  Walter de Jong <walter@heiho.net>
    and Jonathan Niehof <jtniehof@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#include <pwd.h>
#include <time.h>
#include <netdb.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <gdbm.h>

#include "config.h"
#include "pam_shield.h"

#pragma GCC visibility push(hidden)

#define DEFAULT_MAX_CONNS		10
#define DEFAULT_INTERVAL		60L
#define DEFAULT_RETENTION		(3600L * 24L)

#define MAX_LINE				1024

#define OPT_DEBUG				0x001
#define OPT_BLOCK_ALL			0x002		/* block all, including known users */
#define OPT_DRYRUN				0x004
#define OPT_LISTDB				0x008
#define OPT_MISSING_DNS			0x010	/* allow missing DNS */
#define OPT_MISSING_REVERSE		0x020	/* allow missing reverse DNS */
#define OPT_FORCE				0x040	/* purge unexpired entries */
#define OPT_REMOVEIP			0x080
#define OPT_SYNC				0x100

extern int options;
extern GDBM_FILE dbf;

extern char *conffile;
extern char *dbfile;
extern char *trigger_cmd;
extern char *removeip;

/* white lists of addresses */
extern ip_list *allow_ipv4_list;
extern ip_list *allow_ipv6_list;
extern name_list *allow_names;

extern int max_conns;
extern long interval;
extern long retention;

extern time_t this_time;

void logmsg(int level, const char *fmt, ...);

ip_list *new_ip_list(void);

void destroy_ip_list(ip_list *list);

void add_ip_list(ip_list **root, ip_list *ip);

/*
	try to match an IP number against the allow list
	returns 1 if it matches
*/
int match_ipv4_list(unsigned char *saddr);

int match_ipv6_list(unsigned char *saddr);

/*
	name_lists are hostnames and/or network names
*/
name_list *new_name_list(char *name);

void destroy_name_list(name_list *list);

void add_name_list(name_list **root, name_list *n);

/*
	see if 'name' matches our whitelist
	return 1 if it does
*/
int match_name_list(char *name);


/*
	initialize variables
*/
int init_module(void);

void deinit_module(void);

/*
	strip leading and trailing whitespace from a string
*/
void strip(char *str);

/*
	multipliers:
		1s	second
		1m	minute
		1h	hour
		1d	day
		1w	week
		1M	month
		1y	year

	default is 1
	returns 0 on error
*/
long get_multiplier(char *str);

/*
	generate bitmask from '/24' notation

	mask is struct in_addr.saddr, size is the size of the array
	(4 for IPv4, 16 for IPv6)
*/
void ip_bitmask(int bits, unsigned char *mask, int size);

/*
	allow network/netmask, for both IPv4 and IPv6
	netmask can be in canonical or decimal notation
*/
int allow_ip(char *ipnum, int line_no);

/*
	read configuration file
*/
int read_config(void);

/*
	print the IP number of a db_record
	return NULL on error, or buf on success
*/
const char *print_ip(_pam_shield_db_rec_t *record, char *buf, int buflen);

/*
	run external command
*/
int run_trigger(char *cmd, _pam_shield_db_rec_t *record);

int expire_record(_pam_shield_db_rec_t *record);

/*
	gdbm has encountered a fatal error
*/
void fatal_func(const char *str);

#pragma GCC visibility pop
/* EOB */


shield-trigger.sh
#! /bin/sh
#
#	shield-trigger.sh	WJ107
#
#   pam_shield 0.9.4  WJ107
#   Copyright (C) 2007,2010  Walter de Jong <walter@heiho.net>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

null_route() {
#
#	louzy detection of IPv4 or IPv6 address
#
	INET=`echo "$2" | sed 's/[0-9\.]//g'`
	if [ -z "$INET" ]
	then
		INET=""
		GW="127.0.0.1"
	else
		INET="-f inet6"
		GW="::1"
	fi

	if [ -x /sbin/ip ]
	then
		/sbin/ip $INET route $1 blackhole $2
	else
		if [ ! -z "$INET" ]
		then
			INET="-A inet6"
		fi
		/sbin/route $INET $1 -host $2 gw $GW dev lo
	fi

#	mail -s "[security] pam_shield blocked $2" root <<EOF
#Another monkey kept off our backs ...
#EOF
}


### usually no editing is needed beyond this point ###


usage() {
	echo "shield-trigger.sh WJ107"
	echo "usage: ${0##*/} [add|del] <IP number>"
	echo
	echo "shield-trigger.sh is normally called by the pam_shield PAM module"
	exit 1
}


PATH=/sbin:/usr/sbin:/bin:/usr/bin

if [ -z "$2" ]
then
	usage
fi

case "$1" in
	add)
		logger -i -t shield-trigger -p authpriv.info "blocking $2"

		CMD="add"
		IP=$2
		;;

	del)
		logger -i -t shield-trigger -p authpriv.info "unblocking $2"

		CMD="del"
		IP=$2
		;;

	*)
		usage
		;;
esac

null_route "$CMD" "$IP"

exit 0		# make pam_shield happy

# EOB


shield_purge.c
/*
	shield_purge.c

    pam_shield 0.9.7
    Copyright (C) 2007-2012  Walter de Jong <walter@heiho.net>
    and Jonathan Niehof <jtniehof@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "pam_shield.h"

#include <libgen.h>
#include <getopt.h>

#include "pam_shield_lib.h"


void logmsg(int level, const char *fmt, ...) {
va_list varargs;

	if (level == LOG_DEBUG && !(options & OPT_DEBUG))
		return;

	va_start(varargs, fmt);
	vfprintf(stderr, fmt, varargs);
	fprintf(stderr, "\n");
	va_end(varargs);
}


static void usage(char *progname) {
	printf(
"shield-purge " PAM_SHIELD_VERSION "\n"
"usage: %s <options>\n"
"options:\n"
"  -h, --help        Display this information\n"
"  -c, --conf=file   Specify config file (default: " DEFAULT_CONFFILE ")\n"
"  -d, --debug       Verbose output for debugging purposes\n"
"  -n, --dry-run     Do not perform any updates\n"
"  -l, --list        List all database entries\n"
"  -f, --force       Delete all entries, even if unexpired\n"
"  -r, --remove=ip   Remove IP from database\n"
"  -s, --sync        Trigger sync for active records\n"
"                    (rebuild/verify firewall rules)\n"
, basename(progname));

	printf("\n"
"This program is part of the PAM-shield package.\n"
"PAM-shield comes with ABSOLUTELY NO WARRANTY.  This is free software, and you\n"
"are welcome to redistribute it under certain conditions.  See the GNU\n"
"General Public Licence for details.\n"
"\n"
"Copyright (C) 2007-2011 by Walter de Jong <walter@heiho.net>\n"
"Copyright 2010 Jonathan Niehof <jtniehof@gmail.com>\n");
	exit(1);
}

static void get_options(int argc, char **argv) {
int opt;
struct option long_options[] = {
	{ "help",		0, NULL, 'h' },
	{ "debug",		0, NULL, 'd' },
	{ "conf",		1, NULL, 'c' },
	{ "dry-run",	0, NULL, 'n' },
	{ "list",		0, NULL, 'l' },
	{ "force",		0, NULL, 'f' },
	{ "remove",		1, NULL, 'r' },
	{ "sync",		0, NULL, 's' },
	{ NULL,			0, NULL, 0   },
};

	while((opt = getopt_long(argc, argv, "hdc:nlfr:s", long_options, NULL)) != -1) {
		switch(opt) {
			case 'h':
			case '?':
				usage(argv[0]);

			case 'd':
				options |= OPT_DEBUG;
				logmsg(LOG_DEBUG, "logging debug info");
				break;

			case 'c':
				if (optarg == NULL || !*optarg) {
					logmsg(LOG_ERR, "missing filename");
					exit(1);
				}
				if ((conffile = strdup(optarg)) == NULL) {
					logmsg(LOG_ERR, "out of memory");
					exit(-1);
				}
				break;

			case 'n':
				options |= OPT_DRYRUN;
				logmsg(LOG_DEBUG, "performing dry-run");
				break;

			case 'l':
				options |= OPT_LISTDB;
				logmsg(LOG_DEBUG, "list database");
				break;

			case 'f':
				options |= OPT_FORCE;
				logmsg(LOG_DEBUG, "force purge");
				break;

			case 'r':
				options |= OPT_REMOVEIP;
				if (optarg == NULL || !*optarg) {
					logmsg(LOG_ERR, "missing ip");
					exit(1);
				}
				if ((removeip = strdup(optarg)) == NULL) {
					logmsg(LOG_ERR, "out of memory");
					exit(-1);
				}
				break;

			case 's':
				options |= OPT_SYNC;
				logmsg(LOG_DEBUG, "sync");
				break;

			default:
				logmsg(LOG_ERR, "bad command line option");
				exit(1);
		}
	}
}

/*
	lists one record from the DB
*/
static void print_record(_pam_shield_db_rec_t *record) {
char ipbuf[INET6_ADDRSTRLEN];
unsigned int i;
char * time = ctime(&record->trigger_active);

	time[strlen(time)-1] = '\0';

	print_ip(record, ipbuf, INET6_ADDRSTRLEN);

	printf("\n    {\n"
			"      \"ip\": \"%s\",\n"
			"      \"max_entries\": %u,\n"
			"      \"count\": %u,\n"
			"      \"trigger_active\": \"%s\",\n"
			"      \"timestamps\": [\n",
			ipbuf, record->max_entries, record->count,
			(record->trigger_active > 0) ? time : ""
	      );
	for(i = 0; i < record->max_entries; i++)
		if (record->timestamps[i] > 0) {
			time = ctime(&record->timestamps[i]);
			time[strlen(time)-1] = '\0';
			printf("        \"%s\"%s\n", time, (record->timestamps[(i+1)] > 0) ? "," : "");
		}

	printf("      ]\n    }");
}

/*
	list database entries
	this is also mostly meant for debugging and looking at what's in the DB
*/
static void list_db(void) {
_pam_shield_db_rec_t *record;
datum key, next_key, data;

	key = gdbm_firstkey(dbf);

	if (key.dptr == NULL) {
		printf("database is empty\n");
		return;
	}

	printf("{\n  \"db\": [");

	while(key.dptr != NULL) {
		data = gdbm_fetch(dbf, key);

		if (data.dptr != NULL) {
			record = (_pam_shield_db_rec_t *)data.dptr;
			print_record(record);
		}
		next_key = gdbm_nextkey(dbf, key);
		free(key.dptr);
		key = next_key;
		if (data.dptr != NULL && key.dptr != NULL) {
			printf(",");
		}
	}

	printf("\n  ]\n}\n");
}

/*
	expire old entries from the database
*/
static void purge_db(void) {
_pam_shield_db_rec_t *record;
datum key, next_key, data;
int deleted=0; /*If any key deleted, order changes; must revisit all keys*/

	key = gdbm_firstkey(dbf);

	while(key.dptr != NULL) {
		data = gdbm_fetch(dbf, key);
		next_key = gdbm_nextkey(dbf, key);

		if (options & OPT_FORCE) {
			logmsg(LOG_DEBUG, "force-expiring entry");
			if (!(options & OPT_DRYRUN)) {
				gdbm_delete(dbf, key);
				deleted=1;
			}
		}
		else if (data.dptr == NULL) {
			logmsg(LOG_DEBUG, "cleaning up empty key");
			if (!(options & OPT_DRYRUN)) {
				gdbm_delete(dbf, key);
				deleted=1;
			}
		} else {
			record = (_pam_shield_db_rec_t *)data.dptr;

/* store any changes */
			if (expire_record(record)) {
				if (!record->count && !record->trigger_active) {
					logmsg(LOG_DEBUG, "expiring entry");
					if (!(options & OPT_DRYRUN)) {
						gdbm_delete(dbf, key);
						deleted=1;
					}
				} else {
					logmsg(LOG_DEBUG, "storing updated entry");
					if (!(options & OPT_DRYRUN))
						gdbm_store(dbf, key, data, GDBM_REPLACE);
				}
			}
			free(data.dptr);
		}
		free(key.dptr);
		key = next_key;
		if (deleted && !key.dptr) {
			deleted=0;
			key = gdbm_firstkey(dbf);
		}
	}
}

/*
	remove ip from the database
*/
static int remove_ip(void) {
_pam_shield_db_rec_t *record;
datum key, next_key, data;
int deleted=0; /*If any key deleted, order changes; must revisit all keys*/
char ipbuf[INET6_ADDRSTRLEN];

	key = gdbm_firstkey(dbf);

	while(key.dptr != NULL) {
		data = gdbm_fetch(dbf, key);
		next_key = gdbm_nextkey(dbf, key);

		if (data.dptr == NULL) {
			logmsg(LOG_DEBUG, "cleaning up empty key");
			if (!(options & OPT_DRYRUN)) {
				gdbm_delete(dbf, key);
				deleted=1;
			}
		} else {
			record = (_pam_shield_db_rec_t *)data.dptr;

			print_ip(record, ipbuf, INET6_ADDRSTRLEN);
			if (!strcmp(removeip, ipbuf)) {
				logmsg(LOG_DEBUG, "remove entry: %s", ipbuf);
				deleted=1;
				if (!(options & OPT_DRYRUN)) {
					record->trigger_active = (time_t)0L;
					run_trigger("del", record);
					gdbm_delete(dbf, key);
				}
			}
			free(data.dptr);
		}
		free(key.dptr);
		key = next_key;
		if (deleted && !key.dptr) {
			if (!(options & OPT_DRYRUN)) {
				key = gdbm_firstkey(dbf);
			}
			return 0;
		}
	}

	logmsg(LOG_ERR, "not found: %s", removeip);
	return 1;
}

int main(int argc, char **argv) {
	int retval = 0;
	init_module();

	read_config();
	get_options(argc, argv);

	this_time = time(NULL);

	if ((dbf = gdbm_open(dbfile, 512, GDBM_WRITER, (mode_t)0600, fatal_func)) == NULL) {
		logmsg(LOG_ERR, "failed to open db '%s' : %s", dbfile, gdbm_strerror(gdbm_errno));
		return -1;
	}
	if (options & OPT_LISTDB)
		list_db();
	else if (options & OPT_REMOVEIP)
		retval = remove_ip();
	else
		purge_db();

	gdbm_close(dbf);

	deinit_module();
	return retval;
}

/* EOB */
